feat(op, fs): :sparkles: Intergated `statx` into monoio

Lzzzzzt opened the pull request:
# Add `statx` support for monoio

Add `Statx<T>`  struct to support Op, just implement for `SharedFd` and `CString` to support different use cases. 

So there are two types: `FdStatx` and `PathStatx`, both of them are implement `OpAble` Trait.

Currently, they can work on linux with `iouring` and `legacy` feature, and not support macos and windows, due to these two platforms don't have the `statx` syscall, it may use the `stat` syscall and transform the result to the `statx`

Also, to test the `Statx<T>` work correctly, I add some new struct to fs module like `Metadata`, `Permissions` and `FileType`.

There is another reason to add these struct, I can't transform the `statx` struct to the `Metadata` and other types in `std::fs`.

So these struct provide nearly the same API like `std::fs`'s

I also add some basic test in tests dir.

Currently, all these changes above **only** work on linux

Lzzzzzt committed:
feat(op, fs): :sparkles: Intergated `statx` into monoio

Lzzzzzt committed:
feat(op, fs): :sparkles: Intergated `statx` into monoio

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/bytedance/monoio?pullRequest=268) <br/>All committers have signed the CLA.

Lzzzzzt commented:
#267

Lzzzzzt committed:
fix(op, fs): :bug: fix some platform-specific bug

Lzzzzzt committed:
fix(op, fs): :bug: fix some platform-specific bug

Lzzzzzt committed:
docs(fs): update permissions doc

ihciah commented:
It seems test failed for i686 and aarch64...

Lzzzzzt committed:
fix(op, fs): :bug: fix some platform-specific bug

Lzzzzzt commented:
the problem in aarch64 may fixed, but the i686's i don't know why this would happen, maybe disable the statx in i686?

Lzzzzzt committed:
fix(fs): :bug: remove `unwrap` call that will cause panic

Lzzzzzt committed:
feat: :sparkles: make `Statx`(actually is stat) support macos

Lzzzzzt committed:
feat: :sparkles: make `Statx`(actually is stat) support macos

Lzzzzzt committed:
Merge branch 'bytedance:master' into statx

Lzzzzzt committed:
fix(tests): :bug: disable tests on windows

Lzzzzzt committed:
ci: :pushpin: upgrade the cross version with latest git branch

upgrade the cross version with latest git branch, to fix some linker error when CI

Lzzzzzt committed:
Merge branch 'cross-version' into statx

Lzzzzzt commented:
the error happen due to the cross's bug, i have test in newest cross version, it works fine.

Lzzzzzt committed:
refactor: refact the impl structure for different os

Lzzzzzt committed:
cross

Lzzzzzt committed:
Merge branch 'bytedance:master' into statx

Lzzzzzt committed:
refactor: refact the impl structure for different os

Lzzzzzt commented:
<img width="410" alt="image" src="https://github.com/bytedance/monoio/assets/101313294/d9cd72c5-16b9-44fa-9447-aafece7c9bfc">

and the run.sh is
```bash
#!/bin/bash

# Define the list of architectures to test
architectures=(
    "x86_64-unknown-linux-gnu"
    "i686-unknown-linux-gnu"
    "aarch64-unknown-linux-gnu"
    "armv7-unknown-linux-gnueabihf"
    "riscv64gc-unknown-linux-gnu"
    "s390x-unknown-linux-gnu"

    "x86_64-apple-darwin"
    "aarch64-apple-darwin"
)

# Loop through each architecture and run the tests
for arch in "${architectures[@]}"; do

    cross test --lib --bins --tests --target "$arch" --no-default-features --features "async-cancel,bytes,legacy,macros,utils" > /dev/null 2>&1

    if [ $? -eq 0 ]; then
        echo "Success for $arch"
    else
        echo "Failure for $arch"
    fi
done
```

the failure of riscv is [#1423](https://github.com/cross-rs/cross/issues/1423)

ihciah reviewed (approved):
Thank you! I'll merge it first, and I'll look at the testing issue.
